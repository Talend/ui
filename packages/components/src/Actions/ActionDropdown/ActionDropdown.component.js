import PropTypes from 'prop-types';
import React from 'react';
import classNames from 'classnames';
import { DropdownButton, MenuItem, OverlayTrigger } from 'react-bootstrap';
import Inject from '../../Inject';
import theme from './ActionDropdown.scss';
import TooltipTrigger from '../../TooltipTrigger';
import Icon from '../../Icon';
import { wrapOnClick } from '../Action/Action.component';

function InjectDropdownMenuItem({
	getComponent,
	component,
	divider,
	withMenuItem,
	liProps,
	menuItemProps,
	key,
	onSelect,
	onKeyDown,
	...rest
}) {
	const Renderers = Inject.getAll(getComponent, { MenuItem });
	if (divider) {
		return <Renderers.MenuItem key={key} {...menuItemProps} divider />;
	}
	if (withMenuItem) {
		return (
			<Renderers.MenuItem key={key} {...menuItemProps} onSelect={onSelect} onKeyDown={onKeyDown}>
				<Inject component={component} getComponent={getComponent} {...rest} />
			</Renderers.MenuItem>
		);
	}
	return (
		<li role="presentation" key={key} {...liProps}>
			<Inject component={component} getComponent={getComponent} {...rest} />
		</li>
	);
}

InjectDropdownMenuItem.propTypes = {
	getComponent: PropTypes.func.isRequired,
	component: PropTypes.string,
	divider: PropTypes.bool,
	withMenuItem: PropTypes.bool,
	liProps: PropTypes.object,
	menuItemProps: PropTypes.object,
	key: PropTypes.number,
	onSelect: PropTypes.func,
	onKeyDown: PropTypes.func,
};
InjectDropdownMenuItem.displayname = 'InjectDropdownMenuItem';

function getMenuItem(item, index, getComponent) {
	const Renderers = Inject.getAll(getComponent, { MenuItem });
	if (item.divider) {
		return <Renderers.MenuItem key={index} divider />;
	}
	return (
		<Renderers.MenuItem key={index} eventKey={item} {...item} onClick={wrapOnClick(item)}>
			{item.icon && <Icon name={item.icon} />}
			{item.label}
		</Renderers.MenuItem>
	);
}

/**
 * @param {object} props react props
 * @example
 const props = {
	label: 'related items',
	icon: 'fa fa-file-excel-o',
	items: [
		{
			icon: 'talend-icon',
			label: 'document 1',
			onClick: action('document 1 click'),
		},
		{
			divider: true,
		},
		{
			label: 'document 2',
			onClick: action('document 2 click'),
		},
	],
	tooltipPlacement: 'right',
	hideLabel: true,
	link: true,
	onSelect: action('item selected'),
};
 <ActionDropdown {...props} />
 */
function ActionDropdown(props) {
	const {
		bsStyle,
		hideLabel,
		icon,
		items,
		label,
		link,
		onSelect,
		tooltipPlacement,
		tooltipLabel,
		getComponent,
		components,
		...rest
	} = props;

	const Renderers = Inject.getAll(getComponent, { MenuItem, DropdownButton });
	const injected = Inject.all(getComponent, components, InjectDropdownMenuItem);
	const title = (
		<span className="tc-dropdown-button-title">
			{icon ? <Icon name={icon} /> : null}
			{hideLabel ? null : <span className="tc-dropdown-button-title-label">{label}</span>}
		</span>
	);
	const style = link ? 'link' : bsStyle;

	function onItemSelect(object, event) {
		if (onSelect) {
			onSelect(event, object);
		}
	}

	const dropdown = (
		<Renderers.DropdownButton
			title={title}
			bsStyle={style}
			role="button"
			onSelect={onItemSelect}
			className={classNames(theme['tc-dropdown-button'], 'tc-dropdown-button')}
			{...rest}
		>
			{!items.length && !components && <Renderers.MenuItem disabled>No options</Renderers.MenuItem>}
			{injected('beforeItemsDropdown')}
			{items.map((item, key) => getMenuItem(item, key, getComponent))}
			{injected('itemsDropdown')}
			{injected('afterItemsDropdown')}
		</Renderers.DropdownButton>
	);

	if (hideLabel || tooltipLabel) {
		return (
			<TooltipTrigger label={tooltipLabel || label} tooltipPlacement={tooltipPlacement}>
				{dropdown}
			</TooltipTrigger>
		);
	}
	return dropdown;
}

ActionDropdown.displayName = 'ActionDropdown';

ActionDropdown.propTypes = {
	bsStyle: PropTypes.string,
	hideLabel: PropTypes.bool,
	noCaret: PropTypes.bool,
	pullRight: PropTypes.bool,
	icon: PropTypes.string,
	items: PropTypes.arrayOf(
		PropTypes.shape({
			icon: PropTypes.string,
			label: PropTypes.string,
			...MenuItem.propTypes,
		}),
	).isRequired,
	label: PropTypes.string.isRequired,
	link: PropTypes.bool,
	onSelect: PropTypes.func,
	tooltipPlacement: OverlayTrigger.propTypes.placement,
	tooltipLabel: PropTypes.string,
	getComponent: PropTypes.func,
	components: PropTypes.shape({
		beforeItemsDropdown: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
		itemsDropdown: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
		afterItemsDropdown: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
	}),
};

ActionDropdown.defaultProps = {
	bsStyle: 'default',
	tooltipPlacement: 'top',
	items: [],
};

export { ActionDropdown as default, getMenuItem, InjectDropdownMenuItem };
