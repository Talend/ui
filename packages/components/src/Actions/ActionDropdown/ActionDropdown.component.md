# ActionDropdown component

This component display a dropdown with items in it. It used react-bootstrap component [DropdownButton](https://react-bootstrap.github.io/components/dropdowns/#btn-dropdowns), [MenuItem](https://react-bootstrap.github.io/components/dropdowns/#menu-items) and [TooltipTrigger](https://react-bootstrap.github.io/components/tooltips/#tooltips).

A basic example of use

```javascript
function Example() {
	/**
	 * here item will be rendered
	 * as simple label
	 * a list divider
	 * an icon with a simple label
	 */

	const items = [
		{
			label: 'item 1',
		},
		{
			divider: true,
		},
		{
			label: 'item 2',
			icon: 'talend-another-icon-file',
		},
	];
	return (
		<div>
			<ActionDropdown
				id="my-dropdown"
				label="My Dropdown"
				icon="talend-icon-file"
				items={items}
				tooltipLabel="My tooltip label"
				tooltipPlacement="down"
			/>
		</div>
	);
}
```

## How it works

#### 1 ) title

The `title` of the `ActionDropdown` can be parameterized with the followings props:

| Props     | type            |
| --------- | --------------- |
| label     | Required String |
| hidelabel | Bool            |
| icons     | String          |

#### 2 ) items

The list of `items` that the `ActionDropdown` will display.

| Props | type                               |
| ----- | ---------------------------------- |
| items | Required (array OR Immutable.List) |

each composed item being an object composed of

| Properties | type            |
| ---------- | --------------- |
| label      | Required String |
| hidelabel  | Bool            |
| icons      | String          |

or just {divider: true}

```javascript
	items: [
		{
			label: 'my-label-firt-item',
		},
		{
			divider: true,
		},
		{
			label: 'my-label-second-item',
			icon: 'my-icon-second-item',
		},
	],
```

#### 3 ) components

This component support [Inject](../../Inject/Inject.md) api.

Using this props you can define additionnal component that your `ActionDropdown` will show.
(This do no permit to alter the render of `items`)

They are 3 slots in the `ActionDropdown` component where you can inject your custom components.

* `beforeItemsDropdown` will allow you to inject one or many components before the item list
* `itemsDropdown` will allow you to inject one or many components just after the item list
* `afterItemsDropdown` will allow you to inject one or many components before the item list (this should be deprecated since this exactly what `itemsDropdown` will do)

Each of this component should be defined as an object using the following api :

| properties    | type                                                                                       |
| ------------- | ------------------------------------------------------------------------------------------ |
| component     | String (Required if not a divider)                                                         |
| divider       | Bool                                                                                       |
| withMenuItem  | Bool                                                                                       |
| menuItemProps | [react-bootstrap](https://react-bootstrap.github.io/components/dropdowns/#menu-item-props) |
| liProps       | Object                                                                                     |

Your component will be wrapped into a `MenuItem` if `withMenuItem` property is set to true
This will give you access to customisation of [react-bootstrap](https://react-bootstrap.github.io/components/dropdowns/#menu-items) `MenuItem` component

The `<li>` and `<MenuItem>` can be deeply customized with menuItemProps and liProps.
You can add them to your component object

```javascript
components: {
	itemsDropdown: [
		{
			component: 'MyComponent',
			withMenuItem: true,
			menuItemProps: {
				className: 'MySpecialClassName',
				...moreStuff,
			},
		},
	],
}
```

#### 4 ) overlay

You can have a overlay / tooltip on the ActionDropdown.

#### 5 ) select callback

With the onSelect props you can add a callback when an item is selected.

## REF

| Props            | Usage                                                                                                                        |
| ---------------- | ---------------------------------------------------------------------------------------------------------------------------- |
| bsStyle          | string used to define DropdownButton bsStyle (default, primary, sucess, info, warning, danger)                               |
| hideLabel        | boolean that condition the used of overlay                                                                                   |
| icon             | string that defines the icon used in the dropdown button                                                                     |
| items            | array of items displayed in the dropdown                                                                                     |
| label            | string that defines the title used in the dropdown button or in the tooltip                                                  |
| link             | boolean which condition the bsStyle                                                                                          |
| onSelect         | callback used when dropdown clicked                                                                                          |
| tooltipPlacement | string ('up', 'down' ...) to position the tooltip overlay                                                                    |
| tooltipLabel     | string label used to condition the used of overlay and label of overlay                                                      |
| getComponent     | please see the component.md in cmf for more information.                                                                     |
| components       | beforeItemsDropdown, itemsDropdown, afterItemsDropdown : arrays of items or simple object that will be used in the dropdown. |

---

| Props            | default |
| ---------------- | ------- |
| bsStyle          | default |
| tooltipPlacement | top     |
| items            | []      |
