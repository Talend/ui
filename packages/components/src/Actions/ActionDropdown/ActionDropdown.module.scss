@use '~@talend/bootstrap-theme/src/theme/guidelines' as *;
@use '~@talend/design-tokens/lib/tokens';

$tc-dropdown-loader-padding: $padding-small !default;
$tc-dropdown-button-right-padding: 0.5rem;

.tc-dropdown {
	&-button:global(.btn-link) {
		&:hover,
		&:focus,
		&:active {
			text-decoration: none;
		}
	}

	&-button {
		padding-right: $tc-dropdown-button-right-padding;

		.tc-dropdown-caret {
			width: 0.5rem;
			height: 0.5rem;
			transition: transform 0.1s ease-in;
			will-change: transform;

			&.tc-dropdown-caret-open {
				transform: rotate(-180deg);
			}
		}

		&.ellipsis {
			&::after {
				content: '\22ee';
				font-size: 2em;
				font-weight: bold;
				vertical-align: inherit;
			}

			+ ul {
				min-width: auto;
			}
		}
	}

	&-item {
		a img {
			max-width: initial;
		}

		a img,
		a svg {
			margin: 0 $padding-smaller;
		}

		&-badge {
			margin-left: $padding-smaller;
		}
	}

	&-loader {
		margin: -$tc-dropdown-loader-padding 0;
		padding: $tc-dropdown-loader-padding 0;
		text-align: center;
		background: tokens.$coral-color-neutral-background-medium;

		:global(.divider) + & {
			margin: -($tc-dropdown-loader-padding - 1px) 0;
		}
	}
}
