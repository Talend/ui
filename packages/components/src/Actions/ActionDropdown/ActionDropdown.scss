$tc-dropdown-loader-padding: $padding-small !default;

.tc-dropdown {
	&-button:global(.btn-link) {
		&:hover,
		&:focus,
		&:active {
			text-decoration: none;
		}
	}

	&-item {
		a svg {
			margin: 0 $padding-smaller;
		}
	}

	&-loader {
		margin: -$tc-dropdown-loader-padding 0;
		padding: $tc-dropdown-loader-padding 0;
		text-align: center;
		background: $concrete;

		:global(.divider) + & {
			margin: -($tc-dropdown-loader-padding - 1px) 0;
		}
	}
}
