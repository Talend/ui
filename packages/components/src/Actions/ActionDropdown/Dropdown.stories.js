import { action } from '@storybook/addon-actions';
import Immutable from 'immutable';

import FilterBar from '../../FilterBar';
import Action from '../Action';
import ActionDropdown from './ActionDropdown.component';

const myAction = {
	id: 'context-dropdown-related-items',
	label: 'related items',
	icon: 'talend-file-xls-o',
	items: [
		{
			id: 'context-dropdown-item-document-1',
			icon: 'talend-file-json-o',
			label: 'document 1',
			'data-feature': 'actiondropdown.items',
			onClick: action('document 1 click'),
		},
		{
			icon: 'src-data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCAxNiAxNiI+CiAgPGc+CiAgICA8cGF0aCBjbGFzcz0idGktYmFja2dyb3VuZCIgIGZpbGw9IiNGRjZFNzAiIGQ9Ik0xNiw4IEMxNiwxMi40MTg4NzM3IDEyLjQxODI2NDgsMTYuMDAwNjA4OCA4LDE2LjAwMDYwODggQzMuNTgxNzM1MTYsMTYuMDAwNjA4OCAwLDEyLjQxODg3MzcgMCw4IEMwLDMuNTgxNzM1MTYgMy41ODE3MzUxNiw1LjM0OTk2ODgzZS0xNCA4LDUuMzQ5OTY4ODNlLTE0IEMxMi40MTgyNjQ4LDUuMzQ5OTY4ODNlLTE0IDE2LDMuNTgxNzM1MTYgMTYsOCIvPgogICAgPHBhdGggY2xhc3M9InRpLWZvcmVncm91bmQiIGZpbGw9IiNGRkYiIGQ9Ik0xMC4xNzgwMTY4LDExLjQzMTYwMzQgQzEwLjE1MTM5NSwxMS4zNTQxNTggMTAuMTEzODgyNCwxMS4yODIxNTggMTAuMDY0ODczOSwxMS4yMTYyMDg0IEMxMC4wMTY0NzA2LDExLjE0OTY1MzggOS45Nzg5NTc5OCwxMS4xMjMwMzE5IDkuOTUyMzM2MTMsMTEuMTM2MzQyOSBDOS43ODQxMzQ0NSwxMS4yMTk4Mzg3IDkuNjAwMjAxNjgsMTEuMjkxMjMzNiA5LjQwMTE0Mjg2LDExLjM0ODcxMjYgQzkuMjAyMDg0MDMsMTEuNDA2MTkxNiA5LjAyOTA0MjAyLDExLjQzNTIzMzYgOC44ODMyMjY4OSwxMS40MzUyMzM2IEM4LjcxOTI2MDUsMTEuNDM1MjMzNiA4LjU5NDAxNjgxLDExLjM5NzcyMSA4LjUwODEwMDg0LDExLjMyNTcyMSBDOC40MjE1Nzk4MywxMS4yNTI1MTA5IDguMzU5ODY1NTUsMTEuMTE2OTgxNSA4LjMyMjM1Mjk0LDEwLjkyMDM0MjkgQzguMjg0MjM1MjksMTAuNzIzNzA0MiA4LjI2NTQ3ODk5LDEwLjQzNjMwOTIgOC4yNjU0Nzg5OSwxMC4wNTk5NzMxIEw4LjI2NTQ3ODk5LDYuMjQyNzYzMDMgTDkuNjkzMzc4MTUsNi4yNDI3NjMwMyBDOS43NDE3ODE1MSw2LjI0Mjc2MzAzIDkuNzg3MTU5NjYsNi4xOTYxNzQ3OSA5LjgyOTUxMjYxLDYuMTAyMzkzMjggQzkuODcxMjYwNSw2LjAxMDQyNjg5IDkuODkyNDM2OTcsNS44OTU0Njg5MSA5Ljg5MjQzNjk3LDUuNzU4MTI0MzcgQzkuODkyNDM2OTcsNS43MDAwNDAzNCA5Ljg3NzkxNTk3LDUuNjQ4NjExNzYgOS44NDg4NzM5NSw1LjYwNTA0ODc0IEM5LjgyMDQzNjk3LDUuNTYwODgwNjcgOS43ODE3MTQyOSw1LjUzNjY3ODk5IDkuNzMzMzEwOTIsNS41MzI0NDM3IEw4LjI2NTQ3ODk5LDUuNTMyNDQzNyBMOC4yNjU0Nzg5OSwzLjY3OTIgQzguMjY1NDc4OTksMy42MDIzNTk2NiA4LjIxNDY1NTQ2LDMuNTQ3OTA1ODggOC4xMjQ1MDQyLDMuNTEwOTk4MzIgQzcuOTU4MTE3NjUsMy40NTg5NjQ3MSA3Ljg2MDcwNTg4LDMuNjM5ODcyMjcgNy44NjA3MDU4OCwzLjYzOTg3MjI3IEM3LjQxNDE4NDg3LDQuNjcyMDczOTUgNi41Mzc0Nzg5OSw1LjQ2ODkxNDI5IDUuNDUzMjQzNyw1LjgwODM0Mjg2IEM1LjQ1MzI0MzcsNS44MDgzNDI4NiA1LjIzMTc5ODMyLDUuODYwOTgxNTEgNS4yODg2NzIyNyw2LjA5MjcxMjYxIEM1LjI5NTMyNzczLDYuMTEyNjc4OTkgNS4zMDEzNzgxNSw2LjEzMjY0NTM4IDUuMzExMDU4ODIsNi4xNTI2MTE3NiBDNS4zNDE5MTU5Nyw2LjIxNzM1MTI2IDUuMzc5NDI4NTcsNi4yNTA2Mjg1NyA1LjQyMzU5NjY0LDYuMjU2MDczOTUgTDYuMTMzOTE1OTcsNi4yNTYwNzM5NSBMNi4xMzM5MTU5NywxMC42NTE3MDQyIEM2LjEzMzkxNTk3LDExLjI3NTUwMjUgNi4zMjMyOTQxMiwxMS43MTE3Mzc4IDYuNzAyMDUwNDIsMTEuOTU5ODA1IEM3LjA4MDIwMTY4LDEyLjIwNzI2NzIgNy41ODg0MzY5NywxMi4zMzA2OTU4IDguMjI1NTQ2MjIsMTIuMzMwNjk1OCBDOC40MjUyMTAwOCwxMi4zMzA2OTU4IDguNjU2MzM2MTMsMTIuMjkzNzg4MiA4LjkxOTUyOTQxLDEyLjIxODc2MyBDOS4xODI3MjI2OSwxMi4xNDMxMzI4IDkuNDMwNzg5OTIsMTIuMDQ2OTMxMSA5LjY2MzEyNjA1LDExLjkyOTU1MjkgQzkuODk2MDY3MjMsMTEuODEyNzc5OCAxMC4wNjcyOTQxLDExLjY5NDE5MTYgMTAuMTc4MDE2OCwxMS41NzM3ODgyIEMxMC4yMDQ2Mzg3LDExLjU1Njg0NzEgMTAuMjA0NjM4NywxMS41MDkwNDg3IDEwLjE3ODAxNjgsMTEuNDMxNjAzNCIvPgogIDwvZz4KPC9zdmc+Cg==',
			label: 'Button with icon as image',
			onClick: action('Button with icon clicked'),
			type: 'button',
		},
		{
			divider: true,
		},
		{
			id: 'context-dropdown-item-document-2',
			label: 'document 2',
			'data-feature': 'actiondropdown.items',
			onClick: action('document 2 click'),
		},
	],
};

const loadingAdditionalContent = {
	id: 'context-dropdown-related-items',
	label: 'related items',
	loading: true,
	icon: 'talend-file-xls-o',
	items: [],
};

const contentAndLoadingAdditionalContent = {
	...loadingAdditionalContent,
	items: [
		{
			id: 'context-dropdown-item-document-1',
			icon: 'talend-file-json-o',
			label: 'document 1',
			'data-feature': 'actiondropdown.items',
			onClick: action('document 1 click'),
		},
		{
			divider: true,
		},
	],
};

const withImmutable = {
	id: 'context-dropdown-related-items',
	label: 'related immutable items',
	items: Immutable.fromJS([
		{
			id: 'context-dropdown-item-document-1',
			icon: 'talend-file-json-o',
			label: 'document 1',
			'data-feature': 'actiondropdown.items',
			onClick: action('document 1 click'),
		},
		{
			divider: true,
		},
		{
			id: 'context-dropdown-item-document-2',
			label: 'document 2',
			'data-feature': 'actiondropdown.items',
			onClick: action('document 2 click'),
		},
	]),
};

const openWithImmutable = { ...withImmutable, open: true };

const withComponents = {
	id: 'context-dropdown-custom-items',
	label: 'custom items',
	icon: 'talend-file-xls-o',
	getComponent: key => {
		if (key === 'Action') {
			return Action;
		} else if (key === 'FilterBar') {
			return FilterBar;
		}
		throw new Error('Component not found');
	},
	components: {
		itemsDropdown: [
			{
				component: 'Action',
				label: 'First item',
				'data-feature': 'actiondropdown.items',
			},
			{
				divider: true,
			},
			{
				component: 'FilterBar',
				label: 'Second item',
				'data-feature': 'actiondropdown.items',
				onFilter: action('onFilter'),
			},
		],
	},
};

const mixItemsComponents = {
	id: 'context-dropdown-mix-items',
	label: 'mix items',
	getComponent: key => {
		if (key === 'Action') {
			return Action;
		}
		throw new Error('Component not found');
	},
	items: [
		{
			id: 'context-dropdown-item-document-1',
			icon: 'talend-file-json-o',
			label: 'document 1',
			'data-feature': 'actiondropdown.items',
			onClick: action('document 1 click'),
		},
		{
			divider: true,
		},
		{
			id: 'context-dropdown-item-document-2',
			label: 'document 2',
			'data-feature': 'actiondropdown.items',
			onClick: action('document 2 click'),
		},
	],
	components: {
		itemsDropdown: [
			{
				component: 'Action',
				label: 'Third item',
				'data-feature': 'actiondropdown.items',
			},
			{
				divider: true,
			},
			{
				component: 'Action',
				label: 'Fourth item',
				'data-feature': 'actiondropdown.items',
			},
		],
	},
};

const propsTooltip = {
	id: 'context-dropdown-tooltip-items',
	tooltipLabel: 'my tooltip',
	label: 'Tooltip',
	items: [
		{
			id: 'context-dropdown-item-document-1',
			icon: 'talend-file-json-o',
			label: 'document 1',
			'data-feature': 'actiondropdown.items',
			onClick: action('document 1 click'),
		},
		{
			divider: true,
		},
		{
			id: 'context-dropdown-item-document-2',
			label: 'document 2',
			'data-feature': 'actiondropdown.items',
			onClick: action('document 2 click'),
		},
	],
};

const oneEventAction = {
	id: 'context-dropdown-events',
	label: 'Dropdown',
	items: [
		{ id: 'item-1', label: 'Item 1', 'data-feature': 'actiondropdown.items' },
		{
			id: 'item-2',
			label: 'Item 2',
			'data-feature': 'actiondropdown.items',
		},
	],
	onSelect: action('onItemSelect'),
};

export default {
	title: 'Components/Actions/Dropdown',
};

export const Default = () => (
	<div>
		<h3>By default :</h3>
		<div id="default">
			<ActionDropdown {...myAction} />
		</div>
		<h3>With one event handler:</h3>
		<div id="oneEvent">
			<ActionDropdown {...oneEventAction} />
		</div>
		<h3>With hideLabel option</h3>
		<div id="hidelabel">
			<ActionDropdown {...myAction} hideLabel />
		</div>
		<h3>With ellipsis option</h3>
		<div id="ellipsis">
			<ActionDropdown {...myAction} ellipsis />
		</div>
		<h3>Empty option</h3>
		<div id="empty">
			<ActionDropdown {...myAction} items={[]} hideLabel />
		</div>
		<h3>Dropup</h3>
		<div id="dropup">
			<ActionDropdown {...myAction} dropup />
		</div>
		<h3>Automatic Dropup : this is contained in a restricted ".tc-dropdown-container" element.</h3>
		<div
			id="auto-dropup"
			className="tc-dropdown-container"
			style={{ border: '1px solid black', overflow: 'scroll', height: '300px' }}
		>
			Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor
			<br />
			ut labore et dolore magna aliqua.
			<br />
			Ut enim ad minim veniam, quis nostrud exercitation ullamco la
			<br />
			Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor
			<br />
			ut labore et dolore magna aliqua.
			<br />
			Ut enim ad minim veniam, quis nostrud exercitation ullamco la
			<br />
			<br />
			<br />
			<br />
			<p>Scroll me to set overflow on top or down of the container, then open the dropdown.</p>
			<ActionDropdown {...myAction} />
			<br />
			<br />
			<br />
			Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor <br />
			ut labore et dolore magna aliqua.
			<br />
			Ut enim ad minim veniam, quis nostrud exercitation ullamco la Lorem ipsum dolor sit amet,
			consectetur adipiscing elit, sed do eiusmod tempor <br />
			ut labore et dolore magna aliqua.
			<br />
			Ut enim ad minim veniam, quis nostrud exercitation ullamco la
		</div>
		<h3>Type link</h3>
		<div id="typeLink">
			<ActionDropdown {...myAction} link />
		</div>
		<h3>Components Items</h3>
		<div id="withComponents">
			<ActionDropdown {...withComponents} />
		</div>
		<h3>Mix Items</h3>
		<div id="mixComponents">
			<ActionDropdown {...mixItemsComponents} />
		</div>
		<h3>Tool tip</h3>
		<div id="toolTip">
			<ActionDropdown {...propsTooltip} />
		</div>
		<h3>With immutable items :</h3>
		<div id="default">
			<ActionDropdown {...withImmutable} />
		</div>
		<h3>Loading additional content</h3>
		<div id="loadingAdditionalContent">
			<ActionDropdown {...loadingAdditionalContent} />
		</div>
		<h3>Content and loading additional content</h3>
		<div id="contentAndLoadingAdditionalContent">
			<ActionDropdown {...contentAndLoadingAdditionalContent} />
		</div>
		<h3>Opened and with immutable items :</h3>
		<div id="openImmutable">
			<ActionDropdown {...openWithImmutable} />
		</div>
	</div>
);
