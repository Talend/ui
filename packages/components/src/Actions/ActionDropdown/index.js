import ActionDropdown, { DROPDOWN_CONTAINER_CN } from './ActionDropdown.component';

export { DROPDOWN_CONTAINER_CN };
export default ActionDropdown;
