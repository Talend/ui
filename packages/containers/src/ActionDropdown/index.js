import ActionDropdown from './ActionDropdown.connect';

export default ActionDropdown;
