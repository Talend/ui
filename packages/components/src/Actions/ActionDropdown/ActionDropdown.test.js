import React from 'react';
import { mount, shallow } from 'enzyme';
import ActionDropdown, { InjectDropdownMenuItem, getMenuItem } from './ActionDropdown.component';

function getComponent(key) {
	const Fake = props => <div {...props} />;
	Fake.displayName = key;
	return Fake;
}

describe('ActionDropdown', () => {
	it('should call onSelect callback when click on item', () => {
		// given
		const onSelectClick = jest.fn();
		const onItemClick = jest.fn();
		const props = {
			id: 'dropdwon-id',
			label: 'Dropdown',
			onSelect: onSelectClick,
			items: [
				{ id: 'item1', label: 'Item 1', onClick: onItemClick, model: 'model' },
				{ id: 'item2', label: 'Item 2', onClick: onItemClick, model: 'model' },
			],
		};
		const actionDropdownInstance = mount(<ActionDropdown {...props} />);
		const menuItems = actionDropdownInstance.find('MenuItem');

		// when
		menuItems
			.at(0)
			.find('SafeAnchor')
			.simulate('click');

		// then
		expect(onSelectClick).toBeCalledWith(jasmine.anything(), props.items[0]);
		expect(onItemClick.mock.calls[0][1]).toEqual({
			action: { id: 'item1', label: 'Item 1' },
			model: 'model',
		});
		expect(onItemClick.mock.calls[0][0].type).toBe('click');

		// when
		menuItems
			.at(1)
			.find('SafeAnchor')
			.simulate('click');

		// then
		expect(onSelectClick).toBeCalledWith(jasmine.anything(), props.items[1]);
		expect(onItemClick.mock.calls[1][1]).toEqual({
			action: { id: 'item2', label: 'Item 2' },
			model: 'model',
		});
		expect(onItemClick.mock.calls[1][0].type).toBe('click');
	});
});

describe('getMenuItem', () => {
	it('should return a MenuItem with divider', () => {
		expect(getMenuItem({ divider: true })).toMatchSnapshot();
	});
	it('should return a MenuItem with icon and label', () => {
		expect(getMenuItem({ label: 'Toto', icon: 'talend-bell' })).toMatchSnapshot();
	});
	it('should return a MenuItem with label', () => {
		expect(getMenuItem({ label: 'Toto' })).toMatchSnapshot();
	});
});

describe('InjectDropdownMenuItem', () => {
	it('should render MenuItem with props divider', () => {
		const wrapper = shallow(
			<InjectDropdownMenuItem
				getComponent={getComponent}
				key={0}
				menuItemProps={{ stuff: 'MyItemProps' }}
				divider
			/>,
		);
		expect(wrapper.getElement()).toMatchSnapshot();
	});
	it('should render MenuItem with Inject', () => {
		const wrapper = shallow(
			<InjectDropdownMenuItem
				getComponent={getComponent}
				component="Action"
				key={0}
				menuItemProps={{ stuff: 'MyItemProps' }}
				onSelect={jest.fn()}
				onKeyDown={jest.fn()}
				withMenuItem
			/>,
		);
		expect(wrapper.getElement()).toMatchSnapshot();
	});
	it('should render li with Inject', () => {
		const wrapper = shallow(
			<InjectDropdownMenuItem
				getComponent={getComponent}
				component="Action"
				key={0}
				liProps={{ stuff: 'MyLiProps' }}
			/>,
		);
		expect(wrapper.getElement()).toMatchSnapshot();
	});
});
