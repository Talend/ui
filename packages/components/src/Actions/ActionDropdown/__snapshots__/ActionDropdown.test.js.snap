// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InjectDropdownMenuItem should render MenuItem with Inject 1`] = `
<MenuItem
  onKeyDown={[Function]}
  onSelect={[Function]}
  stuff="MyItemProps"
>
  <Inject
    component="Action"
    getComponent={[Function]}
  />
</MenuItem>
`;

exports[`InjectDropdownMenuItem should render MenuItem with props divider 1`] = `
<MenuItem
  divider={true}
  stuff="MyItemProps"
/>
`;

exports[`InjectDropdownMenuItem should render li with Inject 1`] = `
<li
  role="presentation"
  stuff="MyLiProps"
>
  <Inject
    component="Action"
    getComponent={[Function]}
  />
</li>
`;

exports[`getMenuItem should return a MenuItem with divider 1`] = `
<MenuItem
  bsClass="dropdown"
  disabled={false}
  divider={true}
  header={false}
/>
`;

exports[`getMenuItem should return a MenuItem with icon and label 1`] = `
<MenuItem
  bsClass="dropdown"
  disabled={false}
  divider={false}
  eventKey={
    Object {
      "icon": "talend-bell",
      "label": "Toto",
    }
  }
  header={false}
  icon="talend-bell"
  label="Toto"
  onClick={[Function]}
>
  <Icon
    name="talend-bell"
  />
  Toto
</MenuItem>
`;

exports[`getMenuItem should return a MenuItem with label 1`] = `
<MenuItem
  bsClass="dropdown"
  disabled={false}
  divider={false}
  eventKey={
    Object {
      "label": "Toto",
    }
  }
  header={false}
  label="Toto"
  onClick={[Function]}
>
  Toto
</MenuItem>
`;
